psdisk/psdisk.c

```diff
@@ -12,156 +12,570 @@
+		printf("Partition %s has invalid size %s.\n", pNode->pHeader.name, endptr);
+		free(pNode);
+		free(argsStr);
+		return -1;
+	}
+
+	if ((pNode->pHeader.type = strtoul(opts[3], &endptr, 0)) == 0 && strlen(endptr) != 0) {
+		printf("Partition %s has invalid type %s.\n", pNode->pHeader.name, endptr);
+		free(pNode);
+		free(argsStr);
+		return -1;
+	}
+
+	pNode->status = part_save;
+
+	LIST_INSERT_HEAD(&psdisk_common.list, pNode, ptrs);
+	++psdisk_common.pCnt;
+
+	free(argsStr);
 
 	return 0;
 }
 
 
-static int psdisk_parseMemOpt(int optind, char *argv[])
+static int psdisk_parseMem(void)
 {
-	if (argv[optind] == NULL || argv[optind + 1] == NULL) {
-		LOG_ERROR("Invalid number of arguments.");
+	char *argsStr, *endptr;
+	const char *delimiter = ",";
+
+	const int MAX_MEM_OPTS = 2;
+	char *opts[MAX_MEM_OPTS];
+
+	/* Get data from optarg and separete them */
+	if ((argsStr = malloc(strlen(optarg) + 1)) == NULL) {
+		printf("Cannot allocate memory, err: %s.\n", strerror(errno));
+		return -1;
+	}
+
+	strcpy(argsStr, optarg);
+	/* Separate arguments */
+	if ((opts[0] = strtok(argsStr, delimiter)) == NULL || (opts[1] = strtok(NULL, delimiter)) == NULL) {
+		printf("Invalid number of arguments.\n");
+		free(argsStr);
 		return -1;
 	}
 
-	if ((psdisk_common.mem.memSize = strtoul(argv[optind], NULL, 10)) == 0) {
-		LOG_ERROR("Invalid memory size.");
+	if ((psdisk_common.mem.memSize = strtoul(opts[0], &endptr, 0)) == 0 && strlen(endptr) != 0) {
+		printf("Invalid memory size: %s.\n", endptr);
+		free(argsStr);
 		return -1;
 	}
 
-	if ((psdisk_common.mem.sectorSize = strtoul(argv[optind + 1], NULL, 10)) == 0) {
-		LOG_ERROR("Invalid sector size.");
+	if ((psdisk_common.mem.sectorSize = strtoul(opts[1], &endptr, 0)) == 0 && strlen(endptr) != 0) {
+		printf("Invalid sector size: %s.\n", endptr);
+		free(argsStr);
 		return -1;
 	}
 
+	psdisk_common.mem.read = psdisk_readImg;
+	psdisk_common.mem.write = psdisk_writeImg;
+
+	free(argsStr);
+
 	return 0;
 }
 
 
-static int psdisk_createImg(int optind, char *argv[])
+static int psdisk_openImg(char *argv[])
 {
-	int i;
-	char *name;
+	const char* param;
 
-	if ((name = argv[optind]) == NULL) {
-		LOG_ERROR("Invalid number of arguments.");
+	if ((psdisk_common.fileName = argv[1]) == NULL) {
+		printf("Invalid file name.\n");
 		return -1;
 	}
 
-	for (i = 0; i < strlen(name); ++i) {
-		if (!isalnum(name[i]) && name[i] != '/') {
-			LOG_ERROR("Invalid storage name - %s.", name);
-			return -1;
-		}
+	if (psdisk_common.fileName[0] == '-') {
+		printf("First argument has to be a file name.\n");
+		return -1;
 	}
 
-	/* Open/create file in binary mode */
-	if ((psdisk_common.file = fopen(name, "w+b")) == NULL) {
-		LOG_ERROR("Cannot create image file, check path - %s.", name);
+	/* Check whether file exists. */
+	psdisk_common.opts |= (((access(psdisk_common.fileName, F_OK) == 0) && errno == 0) ? 1 : 0) << attr_fileExists;
+
+	if (FILE_EXIST(psdisk_common.opts))
+		param = "r+b";
+	else
+		param = "w+b";
```

Comment:
So we need to manually remove old image before creating new one? Seems like a nuisance

Reply:
If image exists and you want to perform changes, it will be updated. It is one of the functionality which was discussed. If you want to create totally new image with the same name, you are responsible for removing the previous one.